3_harmonize.R

```diff
@@ -1,25 +1,74 @@
 # Source the functions that will be used to build the targets in p3_targets_list
 source("3_harmonize/src/format_columns.R")
 source("3_harmonize/src/clean_wqp_data.R")
+source("3_harmonize/src/clean_conductivity_data.R")
+source("3_harmonize/src/clean_temperature_data.R")
 
 p3_targets_list <- list(
   
-  # All columns in p2_wqp_data_aoi are of class character. Coerce select columns
-  # back to numeric, but first retain original entries in new columns ending in 
-  # "_original". The default option is to format "ResultMeasureValue" and 
-  # "DetectionQuantitationLimitMeasure.MeasureValue" to numeric, but additional
-  # variables can be added using the `vars_to_numeric` argument in format_columns().
-  # By default, format_columns() will retain all columns, but undesired variables 
-  # can also be dropped from the WQP dataset using the optional `drop_vars` argument. 
+  # All columns in p2_wqp_data_aoi are of class character. Coerce select 
+  # columns back to numeric, but first retain original entries in new columns
+  # ending in "_original". The default option is to format "ResultMeasureValue"
+  # and "DetectionQuantitationLimitMeasure.MeasureValue" to numeric, but 
+  # additional variables can be added using the `vars_to_numeric` argument in 
+  # format_columns(). By default, format_columns() will retain all columns, but
+  # undesired variables can also be dropped from the WQP dataset using the 
+  # optional `drop_vars` argument. 
   tar_target(
     p3_wqp_data_aoi_formatted,
     format_columns(p2_wqp_data_aoi)
   ),
   
-  # Harmonize WQP data
+  # Harmonize WQP data by uniting diverse characteristic names under more
+  # commonly-used water quality parameter names, flagging missing records,
+  # and flagging duplicate records. Duplicated rows are identified using 
+  # the argument `duplicate_definition`. By default, a record will be 
+  # considered duplicated if it shares the same organization, site id, date,
+  # time, characteristic name and sample fraction, although a different 
+  # vector of column names can be passed to `clean_wqp_data()` below. By 
+  # default, duplicated rows are flagged and omitted from the dataset. To 
+  # retain duplicate rows, set the argument `remove_duplicated_rows` to FALSE. 
   tar_target(
     p3_wqp_data_aoi_clean,
     clean_wqp_data(p3_wqp_data_aoi_formatted, p1_char_names_crosswalk)
+  ),
+  
+  # Group the WQP data by parameter group in preparation for parameter-specific
+  # data cleaning steps
+  tar_target(
+    p3_wqp_data_aoi_clean_grp,
+    p3_wqp_data_aoi_clean %>%
+      group_by(parameter) %>%
+      tar_group(),
+    iteration = "group"
+  ),
+
+  # Create a table that defines parameter-specific data cleaning functions.
+  tar_target(
+    p3_wqp_param_cleaning_info,
+    tibble(
+      parameter_grp_name = c('conductivity', 'temperature'),
+      cleaning_fxn = c('clean_conductivity_data', 'clean_temperature_data')
```

Comment:
My only current concern is that I don't **_think_** (but I don't actually know) this will be considered a targets dependency if the code within these functions change. Trying to think about how we could get targets to recognize them as function dependencies.

Reply:
EXCELLENT point! Your intuition is correct that the functions are not considered a targets dependency and so changing the function `clean_conductivity_data()` would not trigger a rebuild of `p3_wqp_data_aoi_clean_param`. 

In https://github.com/USGS-R/ds-pipelines-targets-example-wqp/pull/75/commits/f4c335b30c8a717f988453537f61fc5b25989403, I defined the functions as a named list instead of a concatenated string, and then use the names in `fxn_to_use` to feed into `do.call()`. This seems to work - I tried making a trivial change in `clean_conductivity_data()` and the new network graph looks like this:

![print](https://user-images.githubusercontent.com/8785034/180875363-a2ba73c9-7614-43ef-b376-41c4944993ed.png)

The two blue "outdated" shapes in that figure belong to `p3_wqp_param_cleaning_info` and `p3_wqp_data_aoi_clean_param`.

Reply:
☝️ I think there may be interest in this kind of pattern in the future, so it is good to know what options exist within targets for it. 

I'm thinking that because `targets` is R native (yay!), when the script runs 
```r
source("3_harmonize/src/clean_conductivity_data.R")
source("3_harmonize/src/clean_temperature_data.R")
``` 
(at the top of your [3_harmonize](https://github.com/USGS-R/ds-pipelines-targets-example-wqp/commit/f4c335b30c8a717f988453537f61fc5b25989403#diff-fbe19d7c3d7086f36cd528c0695b738010ca34b128b92db3216cfad9371398ebL4))

you add objects of class "function" to the environment that can be accessed. Targets will track those changes appropriately if they are referenced as objects, like you did in your commit to change to a list that had the objects instead of strings for the function names. 

Since this pattern seems well suited for re-use, the one picky thing I'd point out is that the coding is currently redundant (e.g., you have to name the function twice with `list(clean_conductivity_data = clean_conductivity_data)`). I think you need the function name, which is why the current code does this. But perhaps you can instead use a vector of function objects (reverting to closer to what you had, but not turning them into strings) and then you could use `substitute` to extract the function names from the objects, like this:

```r
tar_target(
    p3_wqp_param_cleaning_info,
    tibble(
        parameter_grp_name = c('conductivity', 'temperature'),
        cleaning_fxn = c(clean_conductivity_data, clean_temperature_data)
    )
),

# then in `fxn_to_use()`, you do something like

fxn_to_use <- p3_wqp_param_cleaning_info %>%
        filter(parameter_grp_name == unique(p3_wqp_data_aoi_clean_grp$parameter)) %>%
        pull(cleaning_fxn) # skip `names()` call
        
# ... within the `if {}`
do.call(as.character(substitute(fxn_to_use)), list(wqp_data = p3_wqp_data_aoi_clean_grp))       

```
But `do.call()` also accepts a function object instead of a string so I actually think your current could would work(?) 

`do.call(fxn_to_use, list(wqp_data = p3_wqp_data_aoi_clean_grp))` if `fxn_to_use` is either the string for the function _or_ the function object itself...?

since 
```r
do.call(mean, args = list(x = c(1,2,4,5,6,3,4)))
#and 
do.call("mean", args = list(x = c(1,2,4,5,6,3,4)))
```
do the same thing.